tinywl/tinywl.c

```diff
@@ -102,6 +119,44 @@ struct tinywl_keyboard {
 	struct wl_listener destroy;
 };
 
+static void tinywl_view_destroy(struct tinywl_view *view) {
+	wl_list_remove(&view->map.link);
+	wl_list_remove(&view->unmap.link);
+	wl_list_remove(&view->destroy.link);
+	wl_list_remove(&view->request_move.link);
+	wl_list_remove(&view->request_resize.link);
+	wl_list_remove(&view->request_maximize.link);
+	wl_list_remove(&view->request_fullscreen.link);
+
+	free(view);
+}
+
+static void tinywl_view_addon_handle_destroy(struct wlr_addon *addon) {
+	struct tinywl_view *view = wl_container_of(addon, view, addon);
+
+	tinywl_view_destroy(view);
+}
+static const struct wlr_addon_interface tinywl_view_addon_impl = {
+	.name = "tinywl_view",
+	.destroy = tinywl_view_addon_handle_destroy,
+};
+
+static void tinywl_view_addon_assign(struct tinywl_view *view, struct wlr_addon_set *addons,
+		const void * owner) {
+	wlr_addon_init(&view->addon, addons, owner, &tinywl_view_addon_impl);
+}
+
+static struct tinywl_view *tinywl_view_addon_get(struct wlr_addon_set *addons,
+		const void * owner) {
+	struct wlr_addon *addon =
+		wlr_addon_find(addons, owner, &tinywl_view_addon_impl);
+	if (addon == NULL) {
+		return NULL;
+	}
+	struct tinywl_view *view = wl_container_of(addon, view, addon);
+	return view;
+}
```

Comment:
QQ: where is this being added from?

Reply:
Just some logic to add the view as a addon to the scene_tree so that we're able to get our decoration data. The idea comes from the Sway scene PR :)